Index: src/attrib.c
===================================================================
--- src/attrib.c	(revision 388)
+++ src/attrib.c	(working copy)
@@ -917,7 +917,7 @@
 	}
 	goto Lnodecl;
     }
-    else if (ident == Id::lib)
+    else if (ident == Id::lib || ident == Id::link)
     {
 	if (!args || args->dim != 1)
 	    error("string expected for library name");
@@ -930,14 +930,18 @@
 	    args->data[0] = (void *)e;
 	    if (e->op != TOKstring)
 		error("string expected for library name, not '%s'", e->toChars());
-	    else if (global.params.verbose)
+	    else if (global.params.verbose || ident == Id::link)
 	    {
 		StringExp *se = (StringExp *)e;
 		char *name = (char *)mem.malloc(se->len + 1);
 		memcpy(name, se->string, se->len);
 		name[se->len] = 0;
-		printf("library   %s\n", name);
-		mem.free(name);
+		if (global.params.verbose)
+		    printf("library   %s\n", name);
+		if (ident == Id::link)
+		    global.params.libfiles->push((void *) name);
+		else
+		    mem.free(name);
 	    }
 	}
 	goto Lnodecl;
Index: src/idgen.c
===================================================================
--- src/idgen.c	(revision 388)
+++ src/idgen.c	(working copy)
@@ -216,6 +216,7 @@
     // For pragma's
     { "GNU_asm" },
     { "lib" },
+    { "link" },
     { "msg" },
 
     // For special functions
